src/crypto/public_key/elliptic/curves.js

```diff
@@ -27,107 +27,210 @@
-  return new KeyPair(this.curve, {priv: priv});
+Curve.prototype.keyFromPrivate = function (priv) { // Not for ed25519
+  return new KeyPair(this.curve, { priv: priv });
+};
+
+Curve.prototype.keyFromSecret = function (secret) { // Only for ed25519
+  return new KeyPair(this.curve, { secret: secret });
 };
 
 Curve.prototype.keyFromPublic = function (pub) {
-  return new KeyPair(this.curve, {pub: pub});
+  return new KeyPair(this.curve, { pub: pub });
 };
 
-Curve.prototype.genKeyPair = function () {
-  var r = this.curve.genKeyPair();
-  return new KeyPair(this.curve, {
-    pub: r.getPublic().encode(),
-    priv: r.getPrivate().toArray()
-  });
+Curve.prototype.genKeyPair = async function () {
+  var r, keyPair;
+  if (webCrypto && config.use_native && this.web) {
+    keyPair = await webGenKeyPair(this.name, "ECDSA"); // FIXME is ECDH different?
+  } else if (nodeCrypto && config.use_native && this.node) {
+    keyPair = await nodeGenKeyPair(this.name);
+  } else {
+    var compact = this.curve.curve.type === 'edwards' || this.curve.curve.type === 'mont';
+    r = await this.curve.genKeyPair();
+    if (this.keyType === enums.publicKey.eddsa) {
+      keyPair = { secret: r.getSecret() };
+    } else {
+      keyPair = { pub: r.getPublic('array', compact), priv: r.getPrivate().toArray() };
+    }
+  }
+  return new KeyPair(this.curve, keyPair);
 };
 
 
 function get(oid_or_name) {
-  for (var name in curves) {
-    if (curves[name].oid === oid_or_name || name === oid_or_name) {
-      return new Curve(name, {
-        oid: curves[name].oid,
-        hash: curves[name].hash,
-        cipher: curves[name].cipher,
-        curveName: curves[name].curveName,
-        hashName: curves[name].hashName,
-        nist: curves[name].nist
-      });
+  var name;
+  if (enums.curve[oid_or_name]) {
+    name = enums.write(enums.curve, oid_or_name);
+    return new Curve(name, curves[name]);
+  }
+  for (name in curves) {
+    if (curves[name].oid === oid_or_name) {
+      return new Curve(name, curves[name]);
     }
   }
   throw new Error('Not valid curve');
 }
 
-function generate(curve) {
-  return new Promise(function (resolve) {
-    curve = get(curve);
-    var keyPair = curve.genKeyPair();
-    resolve({
-      oid: curve.oid,
-      Q: new BigInteger(keyPair.getPublic()),
-      d: new BigInteger(keyPair.getPrivate()),
-      hash: curve.hash,
-      cipher: curve.cipher
-    });
-  });
+async function generate(curve) {
+  curve = get(curve);
+  var keyPair = await curve.genKeyPair();
+  return {
+    oid: curve.oid,
+    Q: new BigInteger(keyPair.getPublic()),
+    d: new BigInteger(keyPair.getPrivate()),
+    hash: curve.hash,
+    cipher: curve.cipher
+  };
 }
 
 module.exports = {
   Curve: Curve,
   generate: generate,
   get: get
 };
+
+
+//////////////////////////
+//                      //
+//   Helper functions   //
+//                      //
+//////////////////////////
+
+
+async function webGenKeyPair(name, algorithm) {
```

Comment:
As far as I can tell, the only value for algorithm that you ever pass in is ECDSA, and I believe that webcrypto's generateKey function produces identical results for name = ECDH and name = ECDSA as long as you make the key extractable. So is this ECDH handling code necessary?

Reply:
I assume you're referring to `algorithm`, since `name` just gets the curve name. It's not necessary. I left it there mostly as a reminder of possible values for the `algorithm` field in `generateKey`, which are `deriveKey/deriveBits` or `sign/verify`. I couldn't find a nice and clean way of using `deriveKey` in `ecdsa.js`, but if you can we should use it, and in that case, this `algorithm` input could be used.

Reply:
Then for simplicity's sake I think we should get rid of the algorithm parameter and the separate ECDH handling altogether rather than keeping this dead code around (which will probably confuse someone else at some point too).

Reply:
Will do.